File: alligator/src/main/java/me/aartikov/alligator/ScreenSwitcher.java
package me.aartikov.alligator;

/**
 * Date: 22.01.2017
 * Time: 22:46
 *
 * @author Artur Artikov
 */

import android.support.annotation.Nullable;

import me.aartikov.alligator.exceptions.ScreenSwitchingException;

/**
 * Object for switching between several screens without theirs recreation.
 */
public interface ScreenSwitcher {
	/**
	 * Switches to a given screen.
	 *
	 * @param screen        screen
	 * @param animationData animation data for an additional animation configuring
	 * @throws me.aartikov.alligator.exceptions.ScreenSwitchingException on fail
	 */
	void switchTo(Screen screen, @Nullable AnimationData animationData) throws ScreenSwitchingException;

	/**
	 * Returns a current screen.
	 *
	 * @return current screen or {@code null} if there is no current screen.
	 */
	@Nullable
	Screen getCurrentScreen();
}

File: alligator/src/main/java/me/aartikov/alligator/screenswitchers/ScreenSwitcher.kt
package me.aartikov.alligator.screenswitchers

import me.aartikov.alligator.Screen
import me.aartikov.alligator.animations.AnimationData
import me.aartikov.alligator.exceptions.NavigationException
import me.aartikov.alligator.listeners.ScreenSwitchingListener

/**
 * Object for switching between several screens without theirs recreation.
 */
interface ScreenSwitcher {
    /**
     * Switches to a given screen. Implementation must call the passed screen switching listener properly.
     *
     * @param screen        screen
     * @param listener      screen switching listener
     * @param animationData animation data for an additional animation configuring
     * @throws me.aartikov.alligator.exceptions.NavigationException on fail
     */
    @Throws(NavigationException::class)
    fun switchTo(screen: Screen?, listener: ScreenSwitchingListener?, animationData: AnimationData?)
}